### PR TITLE
[CUDA] Support user compute stream with CUDA graph in CUDA plugin EP

### PR DESCRIPTION
**Pull request by tianleiwu**
## Description

The CUDA plugin EP previously rejected combining a user-provided compute stream
(`user_compute_stream`) with CUDA graph capture (`enable_cuda_graph`), returning
`ORT_INVALID_ARGUMENT`. This PR removes that restriction so the two options can
be used together: when both are set, graph capture and replay run on the
user-owned stream (the same stream the kernels are issued to), matching the
bundled (non-plugin) CUDA EP behavior. Several supporting fixes make capture on a
shared stream stable and Memcpy-free.

### Allow user stream + CUDA graph

| File | Change |
|------|--------|
| [onnxruntime/core/providers/cuda/plugin/cuda_ep_factory.cc](onnxruntime/core/providers/cuda/plugin/cuda_ep_factory.cc) | Remove the validation that rejected `user_compute_stream` + `enable_cuda_graph` together. |
| [onnxruntime/core/providers/cuda/plugin/cuda_ep.cc](onnxruntime/core/providers/cuda/plugin/cuda_ep.cc) | `PerThreadContext` accepts an optional external graph stream. When both options are set it captures/replays on the user stream and does **not** create or destroy it (the user owns its lifetime); otherwise it owns a dedicated graph stream as before. |

### Stable, Memcpy-free CUDA graph capture

| File | Change |
|------|--------|
| [onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h](onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h) | Route kernel scratch/workspace allocations through the EP allocator (BFC arena) instead of raw `cudaMallocAsync`/`cudaMalloc`. After warmup the arena reaches steady state, so the capture run serves scratch from already-reserved chunks and the device free-memory footprint stays stable — required for correct capture. Matches the built-in CUDA EP. |
| [onnxruntime/core/providers/cuda/tensor/shape_op.cc](onnxruntime/core/providers/cuda/tensor/shape_op.cc) | Add an adapter-based `Shape` kernel under `#ifdef BUILD_CUDA_EP_AS_PLUGIN` with identical semantics to the CPU `Shape`. Registering `Shape` on the EP keeps it off the CPU EP and avoids the Memcpy nodes that would otherwise break CUDA graph capture. |
| [cmake/onnxruntime_providers_cuda_plugin.cmake](cmake/onnxruntime_providers_cuda_plugin.cmake) | Stop excluding `shape_op.cc` from the plugin build so the adapter-based `Shape` kernel is compiled in. |

### Null-allocator fallback in PrePack (plugin boundary)

In the plugin build the `AllocatorPtr` passed to `PrePack` can arrive null across
the library boundary. Each kernel now falls back to its own default-memory
allocator (`Info().GetAllocator(OrtMemTypeDefault)`), which is always valid.

- [onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc](onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc)
- [onnxruntime/contrib_ops/cuda/moe/moe_quantization.cc](onnxruntime/contrib_ops/cuda/moe/moe_quantization.cc)
- [onnxruntime/contrib_ops/cuda/quantization/matmul_nbits.cc](onnxruntime/contrib_ops/cuda/quantization/matmul_nbits.cc)

### Misc

- [onnxruntime/core/framework/session_state.cc](onnxruntime/core/framework/session_state.cc) — wrap a long line (no behavior change).

## Testing

- New test: [onnxruntime/test/providers/cuda/plugin/cuda_plugin_user_stream_graph_test.cc](onnxruntime/test/providers/cuda/plugin/cuda_plugin_user_stream_graph_test.cc) covering:
  1. Session creation succeeds with both `user_compute_stream` and `enable_cuda_graph` set (regression for the removed validation).
  2. Capture + replay on the user stream produce correct results.
  3. Replay after an in-place input update on the user stream is correct.
- Tests are gated on `ORT_UNIT_TEST_HAS_CUDA_PLUGIN_EP` and skip gracefully when no CUDA device or plugin library is available.

## Motivation and Context

Users that drive ORT from their own CUDA stream (e.g. to interleave ORT inference
with their own kernels) previously could not also benefit from CUDA graph capture
on the plugin EP. This change brings the plugin EP to parity with the bundled
CUDA EP for that workflow.

## Checklist

- [x] Tests added/updated
- [x] No breaking changes (relaxes a previously rejected option combination)
- [ ] Documentation updated (if applicable)

### CODE REVIEW

**Comment by yuslepukhin** [comment]
>   // Keep is_packed=false so the original fp16/bf16 head_sink remains available to the Flash/fallback

The pre-packing algo expects the allocator to be CPU based and in some cases attempts to externalize them to disk. #Resolved

---
Refers to: onnxruntime/contrib_ops/cuda/bert/group_query_attention.cc:144 in 789c771. [](commit_id = 789c771276a260b8570ac4374608c5b29bf88afb, deletion_comment = False)

File: onnxruntime/core/providers/cuda/tensor/shape_op.cc [hunk 1]
```diff
@@ -2,12 +2,83 @@
...
 namespace onnxruntime {
 namespace cuda {
 
+#ifdef BUILD_CUDA_EP_AS_PLUGIN
+// The bundled CUDA EP registers the CPU `onnxruntime::Shape` kernel (which
+// derives from the framework `OpKernel`) and only marks its output as CPU
+// memory. That class cannot be registered through the plugin EP's adapter
+// kernel machinery, so the plugin build provides an adapter-based Shape kernel
+// with identical semantics. Registering Shape on the EP (rather than letting it
+// fall back to the CPU EP) avoids the Memcpy nodes that an isolated CPU Shape
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/tensor/shape_op.cc]: >avoids the Memcpy nodes [](http://example.com/codeflow?start=28&length=23)

The output is still copied so copies are partially avoided.

File: onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h [hunk 2]
```diff
@@ -1023,56 +1023,29 @@ class CudaKernel : public OpKernel {
...
+    //
+    // The previous behavior (cudaMallocAsync/cudaMalloc allocated-and-freed per
+    // call) allocated new device memory on every run, including the capture run,
+    // so no amount of warmup could stabilize it and the
+    // "GPU memory was allocated during CUDA graph capture" detector would trip.
+    // This now matches the built-in (non-plugin) CUDA EP, which also obtains
+    // scratch from Info().GetAllocator() (see core/providers/cuda/cuda_kernel.h).
+    return ::onnxruntime::IAllocator::MakeUniquePtr<T>(
+        Info().GetAllocator(OrtMemType::OrtMemTypeDefault), cnt, /*use_reserve*/ false,
+        /*stream*/ nullptr);
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h]: Yes, this is an issue if CudaMempool arena is used which is truly stream oriented.

File: onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h [hunk 3]
```diff
@@ -1023,56 +1023,29 @@ class CudaKernel : public OpKernel {
   template <typename T>
   using IAllocatorUniquePtr = std::unique_ptr<T, std::function<void(T*)>>;
   template <typename T>
-  inline IAllocatorUniquePtr<T> GetScratchBuffer(size_t cnt, void* s) const {
+  inline IAllocatorUniquePtr<T> GetScratchBuffer(size_t cnt, void* /*s*/) const {
     if (cnt == 0) return IAllocatorUniquePtr<T>(nullptr, [](T*) {});
-    size_t sz = 0;
-    if (!detail::TryBytesForCount(cnt, detail::SizeOf<T>::value, sz)) {
-      ORT_THROW("CUDA scratch buffer allocation size overflow for ", cnt, " elements");
-    }
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h:1031]: No overflow check removed — REGRESSION
The old GetScratchBuffer had:
```.cpp
size_t sz = 0;
if (!detail::TryBytesForCount(cnt, detail::SizeOf<T>::value, sz)) {
    ORT_THROW("CUDA scratch buffer allocation size overflow for ", cnt, " elements");
}
```

File: onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h]: Fixed in 6bfd824. `GetScratchBuffer` now forwards the compute stream to `IAllocator::MakeUniquePtr` (wrapping the raw `cudaStream_t` in `OrtStreamAdapter`/`PluginStreamShim`), so the allocation and any stream-aware free are ordered on the same stream the kernel enqueues work on, matching the built-in CUDA EP and avoiding the mempool-arena race. The stream is only consulted when the allocator reports `IsStreamAware()`, so the default BFC-arena path and CUDA-graph capture stability are unchanged. One caveat: in NVCC (`.cu`) translation units the framework `Stream` type is incomplete and the shim cannot be constructed, so those call sites still pass a null stream; the scratch they request is short-lived and freed within the same Run.

File: onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h (hunk 3 above)
Comment (tianleiwu) [reply onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h:1031]: The overflow check is preserved, it just moved into `IAllocator::MakeUniquePtr<T>`, which calls `ValidatedCalcMemSizeForArray(cnt, sizeof(T))` and throws (`Invalid size requested for allocation`) on `cnt * sizeof(T)` overflow before any allocation happens. This matches the built-in CUDA EP `GetScratchBuffer` (core/providers/cuda/cuda_kernel.h), which likewise relies on `MakeUniquePtr` for the check rather than a hand-rolled `TryBytesForCount`. For the `GetScratchBuffer<void>(bytes, ...)` callers the value is treated as a byte count with no multiplication, so there is nothing to overflow. I added a comment in 6bfd824 noting where the check now lives.

File: onnxruntime/core/providers/cuda/tensor/shape_op.cc (hunk 1 above)
Comment (tianleiwu) [reply onnxruntime/core/providers/cuda/tensor/shape_op.cc]: Good point, updated the comment in 6bfd824 to be accurate: `Shape` reads only the input shape metadata (never its data) and writes the dims to a CPU output, so registering it on the CUDA EP keeps the node inside the device partition and avoids the device->host Memcpy node the framework would otherwise insert to feed an isolated CPU `Shape` node (that Memcpy is what breaks CUDA-graph capture). The output stays in CPU memory, so a downstream device consumer may still need a copy; copies are reduced, not fully eliminated.

**Comment by tianleiwu** [comment]
> The pre-packing algo expects the allocator to be CPU based and in some cases attempts to externalize them to disk.

Re: `group_query_attention.cc` PrePack head_sink fallback. This kernel keeps `is_packed = false` and never populates `prepacked_weights`, so the framework does not register a `PrePackedWeights` container for this input and there is no disk externalization of the result. The allocator is used only to materialize the cached FP32 attention sink in device memory (an H2D copy followed by a conversion kernel), so a default-memory (device) allocator is exactly the kind required here, and a CPU allocator would be wrong. I added a comment clarifying this in 6bfd824. The same reasoning applies to the QMoE and MatMulNBits PrePack fallbacks, which also keep `is_packed = false`.

**Review by yuslepukhin — DISMISSED** [review]
:shipit:

File: onnxruntime/core/providers/cuda/plugin/cuda_ep.cc [hunk 4]
```diff
@@ -83,20 +83,29 @@ void DestroyCudaStreamForDevice(cudaStream_t stream, int device_id) {
...
+  // When external_graph_stream is non-null (user_compute_stream combined with CUDA graph),
+  // capture and replay happen on that user-owned stream so they see the same stream as the
+  // kernels; the context neither creates nor destroys it. Otherwise the context creates and
+  // owns a dedicated graph stream.
+  explicit PerThreadContext(int device_id, cudaStream_t external_graph_stream = nullptr)
       : device_id(device_id),
-        graph_stream(CreateCudaStreamForDevice(device_id)),
+        owns_graph_stream(external_graph_stream == nullptr),
+        graph_stream(external_graph_stream != nullptr ? external_graph_stream
+                                                      : CreateCudaStreamForDevice(device_id)),
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/core/providers/cuda/plugin/cuda_ep.cc]: Fixed in 968fbe87d8. `PerThreadContext` now takes an explicit `bool use_external_stream` and derives ownership from it (`owns_graph_stream = !use_external_stream`) instead of from whether the handle is null. `GetPerThreadContext()` sets `use_external_stream = has_user_compute_stream && enable_cuda_graph` and passes the user's stream — which may legitimately be `cudaStream_t(0)`/nullptr (the CUDA default stream) — as `external_stream`. So selecting the default stream as the user compute stream while enabling CUDA graph now captures/replays on that user-selected stream, and the context never creates or destroys it.

File: onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h [hunk 5]
```diff
@@ -1025,54 +1025,54 @@ class CudaKernel : public OpKernel {
...
+    // the built-in CUDA EP, which forwards its compute stream to MakeUniquePtr.
+    // The overflow check that the previous hand-rolled path performed is still
+    // enforced inside MakeUniquePtr via ValidatedCalcMemSizeForArray (it throws
+    // on cnt * sizeof(T) overflow).
+#ifndef __CUDACC__
+    onnxruntime::OrtStreamAdapter ort_stream(s);
+    return ::onnxruntime::IAllocator::MakeUniquePtr<T>(
+        Info().GetAllocator(OrtMemType::OrtMemTypeDefault), cnt, /*use_reserve*/ false,
+        ort_stream.get());
+#else
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h]: Fixed in 968fbe87d8. `GetScratchBuffer` no longer synthesizes a temporary `PluginStreamShim`/`OrtStreamAdapter` and forwards its `Stream*` to the allocator. You're right that the stream-aware arena persists the pointer (`chunk->stream`) and later dereferences it through the EP stream API (`SyncStream_GetImpl`/`SyncStream_GetSyncId`), so a stack temporary would dangle. It would also be type-confused: a plugin kernel only has the raw `cudaStream_t` (`KernelContext::GetGPUComputeStream`), never the real `OrtSyncStream*` the arena expects, and there is no stable framework `Stream*`/`OrtSyncStream*` available at the plugin kernel layer to hand it instead.

So `GetScratchBuffer` now passes a null stream. The arena then tracks scratch chunks as freely reusable (the same semantics as a plain non-stream-aware BFC arena). CUDA-graph capture stability is preserved because it comes from serving scratch out of already-reserved arena chunks, not from stream tagging, and the graph path runs on a single unified stream. This supersedes my earlier stream-forwarding change in 6bfd824.

File: onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h]: Update: the stream-forwarding approach from 6bfd824 is replaced in 968fbe87d8. As the follow-up review pointed out, wrapping the raw `cudaStream_t` in a stack `PluginStreamShim` and handing its `Stream*` to the stream-aware arena left a dangling/type-confused pointer — the arena persists `chunk->stream` and later dereferences it through the EP stream API (`SyncStream_GetImpl`/`SyncStream_GetSyncId`), and a plugin kernel only exposes the raw `cudaStream_t`, not a real `OrtSyncStream*`.

`GetScratchBuffer` now passes a null stream. For the default BFC plugin arena this is safe (chunks become freely reusable, same as a non-stream-aware arena) and capture stability still holds via arena chunk reuse. The CudaMempool concern raised earlier would require a real long-lived `OrtSyncStream*` plumbed down to the plugin kernel layer, which isn't available today; the plugin forces a single unified stream on these paths.

File: onnxruntime/core/providers/cuda/plugin/cuda_ep.cc [hunk 6]
```diff
@@ -392,7 +405,11 @@ OrtStatus* ORT_API_CALL CudaEp::CreateSyncStreamForDeviceImpl(
   auto cuda_stream = std::make_unique<CudaSyncStream>(ep->factory_, device_id, this_ptr);
 
   if (ep->config_.has_user_compute_stream && ep->config_.user_compute_stream != nullptr) {
-    // Wrap the user-provided external CUDA stream with full cuBLAS/cuDNN handles.
+    // A user-provided compute stream is honored for kernels regardless of whether CUDA graph
+    // capture is enabled — this branch is taken in both graph and non-graph runs. Wrap the
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/core/providers/cuda/plugin/cuda_ep.cc]: Fixed in 45b1e11f80. `CreateSyncStreamForDeviceImpl` now uses `has_user_compute_stream` as the source of truth instead of checking whether `user_compute_stream` is non-null, so `cudaStream_t(0)`/nullptr is treated as an explicit user-selected CUDA default stream in both graph and non-graph runs.

File: onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h [hunk 7]
```diff
@@ -1023,56 +1023,60 @@ class CudaKernel : public OpKernel {
...
-        // Fall back to synchronous free if async free is unsupported or if the
-        // stream is no longer registered. cudaFree is valid for allocations
-        // returned by cudaMallocAsync and avoids using a stale stream handle.
-        cudaFree(ptr);
-      }
-    });
+    // Route kernel scratch/workspace allocations through the EP allocator
+    // (a BFC arena by default) instead of raw cudaMallocAsync/cudaMalloc.
+    //
+    // The arena pre-reserves device memory and reuses freed chunks across runs.
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/core/providers/cuda/plugin/cuda_kernel_adapter.h:1032]: Fixed in 3ce7ff3633. Since `GetScratchBuffer` cannot safely provide a framework `OrtSyncStream*` today and therefore null-tags scratch chunks, `CudaEp::IsConcurrentRunSupportedImpl()` now returns false. That prevents overlapping runs on different CUDA streams from reusing a scratch chunk while earlier work is still in flight. The comments/docs now call out that re-enabling concurrent multi-stream runs requires future C-API support to expose a stable framework stream or sync id to plugin kernels.

File: docs/cuda_plugin_ep/cuda_plugin_ep_design.md [hunk 8]
```diff
@@ -459,9 +459,18 @@ The NHWC rollout is effectively in a "runtime enabled, cleanup remaining" state:
...
+
+The enabling adapter change is in [`include/onnxruntime/ep/adapter/allocator.h`](../../include/onnxruntime/ep/adapter/allocator.h): `IAllocatorWrappingOrtAllocator` now implements `IsStreamAware()`/`AllocOnStream()` (previously `ORT_NOT_IMPLEMENTED`) by forwarding to the underlying `OrtAllocator`'s `AllocOnStream` when it is available (ORT \u2265 1.23), falling back to plain `Alloc` otherwise. This is what lets the forwarded scratch-buffer stream produce a stream-ordered allocation.
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply docs/cuda_plugin_ep/cuda_plugin_ep_design.md]: Fixed in 3ce7ff3633. The design doc no longer says `GetScratchBuffer` forwards the compute stream or produces stream-ordered scratch allocations. It now matches the implementation: scratch uses the EP arena with a null stream tag because plugin kernels only have the raw `cudaStream_t`, and stream-tagged scratch allocation is listed as future C-API work.

File: .github/workflows/linux_cuda_plugin_ci.yml [hunk 9]
```diff
@@ -141,3 +141,27 @@ jobs:
               cd /onnxruntime_src/onnxruntime/test/python/transformers
               python test_cuda_plugin_ep.py
             "
+
+      # --- Run the CUDA plugin EP C++ GoogleTest binary ---
+      # onnxruntime_provider_test is built into the artifact and links the plugin tests
+      # (gated by ORT_UNIT_TEST_HAS_CUDA_PLUGIN_EP). The user-stream + CUDA graph test
+      # registers the plugin .so via GetSharedLibraryFileName("onnxruntime_providers_cuda_plugin"),
+      # which is resolved relative to the binary directory, so the test must run from
+      # /build/Release/Release where the .so lives.
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply .github/workflows/linux_cuda_plugin_ci.yml]: Fixed in a1a2574c9d. The workflow comment now says `GetSharedLibraryFileName(...)` returns only the platform-specific filename, so the C++ test runs from `/build/Release/Release` to make that filename resolve to the built plugin `.so`.

**Comment by yuslepukhin** [comment]
## Associated Issues / Duplicate Check ##
- No linked GitHub issue in the PR description.
- Follow-up PR #29244 (stacked on this) properly exposes KernelContext_GetSyncStream to solve the null-stream-tag limitation for concurrent runs.


## Missing negative tests ##

- No test verifying that user_compute_stream + external allocator still correctly rejects
- No test for cudaStream_t(0) as the user stream (edge case explicitly discussed in code)
- No test for graph annotation ID switching with user stream

## Verdict ##
The PR is functionally correct, well-documented, and conservative in its safety guarantees (disabling concurrent runs until the proper API exists). The code changes are minimal and well-targeted. The main concern is minor missing negative test coverage. The review comments were adequately addressed. Ready for merge pending the final approval from yuslepukhin (who already approved an earlier revision).

**Review by yuslepukhin — APPROVED** [review]
